File: src/parser/nim/parser/parsers/option.nim
import std/tables

import ../helpers/[tree_add, types, charsets]
import ../helpers/[error, validate, forward, rollback, brace_checks]

# ------------------------------------------------------------ Parsing Breakdown
# - value
#  |     ^-EOL-Whitespace-Boundary 2
#  ^-Whitespace-Boundary 1
# ^-Bullet
#  ^-Value
# ------------------------------------------------------------------------------
#
# @param  {object} S - State object.
# @return {undefined} - Nothing is returned.
proc p_option*(S: State): Node =
    let text = S.text
    var state = Bullet
    var `type` = "escaped"
    var N = node(nkOption, S)
    var qchar: char
    var comment = false
    var braces: seq[int] = @[]

    # Error if flag scope doesn't exist.
    bracechecks(S, check = "pre-existing-fs")

    let l = S.l; var c, p: char
    while S.i < l:
        p = c
        c = text[S.i]

        if c in C_NL:
            rollback(S)
            N.stop = S.i
            break # Stop at nl char.

        if c == C_NUMSIGN and p != C_ESCAPE and (state != Value or comment):
            rollback(S)
            N.stop = S.i
            break

        case (state):
            of Bullet:
                N.bullet.start = S.i
                N.bullet.stop = S.i
                N.bullet.value = $c
                state = Spacer

            of Spacer:
                if c notin C_SPACES: error(S, currentSourcePath)
                state = WsbPrevalue

            of WsbPrevalue:
                if c notin C_SPACES:
                    rollback(S)
                    state = Value

            of Value:
                if N.value.value == "":
                    # Determine value type.
                    if c == C_DOLLARSIGN: `type` = "command-flag"
                    elif c == C_LPAREN:
                        `type` = "list"
                        braces.add(S.i)
                    elif c in C_QUOTES:
                        `type` = "quoted"
                        qchar = c

                    N.value.start = S.i
                    N.value.stop = S.i
                    N.value.value = $c
                else:
                    case `type`:
                        of "escaped":
                            if c in C_SPACES and p != C_ESCAPE:
                                state = EolWsb
                                forward(S)
                                continue
                        of "quoted":
                            if c == qchar and p != C_ESCAPE:
                                state = EolWsb
                            elif c == C_NUMSIGN and qchar == C_NULLB:
                                comment = true
                                rollback(S)
                        else: # list|command-flag
                            # The following character after the initial
                            # '$' must be a '('. If it does not follow,
                            # error.
                            #   --help=$"cat ~/files.text"
                            #   --------^ Missing '(' after '$'.
                            if `type` == "command-flag":
                                if N.value.value.len == 1 and c != C_LPAREN:
                                    error(S, currentSourcePath)

                            # The following logic, is precursor validation
                            # logic that ensures braces are balanced and
                            # detects inline comment.
                            if p != C_ESCAPE:
                                if c == C_LPAREN and qchar == C_NULLB:
                                    braces.add(S.i)
                                elif c == C_RPAREN and qchar == C_NULLB:
                                    # If braces len is negative, opening
                                    # braces were never introduced so
                                    # current closing brace is invalid.
                                    if braces.len == 0: error(S, currentSourcePath)
                                    discard braces.pop()
                                    if braces.len == 0:
                                        state = EolWsb

                                if c in C_QUOTES:
                                    if qchar == C_NULLB:
                                        qchar = c
                                    elif qchar == c:
                                        qchar = C_NULLB

                                if c == C_NUMSIGN and qchar == C_NULLB:
                                    if braces.len == 0:
                                        comment = true
                                        rollback(S)
                                    else:
                                        S.column = braces.pop() - S.tables.linestarts[S.line]
                                        inc(S.column) # Add 1 to account for 0 base indexing.
                                        error(S, currentSourcePath)

                    N.value.stop = S.i
                    N.value.value &= $c

            of EolWsb:
                if c notin C_SPACES: error(S, currentSourcePath)

            else: discard

        forward(S)

    discard validate(S, N)
    add(S, N)

File: src/parser/js1/parsers/option.js
"use strict";

const node = require("../helpers/nodes.js");
const error = require("../helpers/error.js");
const add = require("../helpers/tree-add.js");
const { nk } = require("../helpers/enums.js");
const rollback = require("../helpers/rollback.js");
const validate = require("../helpers/validate.js");
const bracechecks = require("../helpers/brace-checks.js");
const {
	cin,
	cnotin,
	C_NL,
	C_SPACES,
	C_QUOTES
} = require("../helpers/charsets.js");

/**
 * ----------------------------------------------------------- Parsing Breakdown
 * - value
 *  |     ^-EOL-Whitespace-Boundary 2
 *  ^-Whitespace-Boundary 1
 * ^-Bullet
 *  ^-Value
 * -----------------------------------------------------------------------------
 *
 * @param  {object} S - State object.
 * @return {undefined} - Nothing is returned.
 */
module.exports = (S) => {
	let l = S.l;
	let state = "bullet";
	let type = "escaped";
	let N = node(nk.Option, S);
	let qchar = "";
	let comment = false;
	let braces = [];

	// Error if flag scope doesn't exist.
	bracechecks(S, null, "pre-existing-fs");

	let c,
		p = "";
	for (; S.i < l; S.i++, S.column++) {
		p = c;
		c = S.text.charAt(S.i);

		if (cin(C_NL, c)) {
			N.end = rollback(S) && S.i;
			break; // Stop at nl char.
		}

		if (c === "#" && p !== "\\" && (state !== "value" || comment)) {
			rollback(S);
			N.end = S.i;
			break;
		}

		switch (state) {
			case "bullet":
				N.bullet.start = N.bullet.end = S.i;
				N.bullet.value = c;
				state = "spacer";

				break;

			case "spacer":
				if (cnotin(C_SPACES, c)) error(S);
				state = "wsb-prevalue";

				break;

			case "wsb-prevalue":
				if (cnotin(C_SPACES, c)) {
					rollback(S);
					state = "value";
				}

				break;

			case "value":
				{
					if (!N.value.value) {
						// Determine value type.
						if (c === "$") type = "command-flag";
						else if (c === "(") {
							type = "list";
							braces.push(S.i);
						} else if (cin(C_QUOTES, c)) {
							type = "quoted";
							qchar = c;
						}

						N.value.start = N.value.end = S.i;
						N.value.value = c;
					} else {
						switch (type) {
							case "escaped":
								if (cin(C_SPACES, c) && p !== "\\") {
									state = "eol-wsb";
									continue;
								}

								break;

							case "quoted":
								if (c === qchar && p !== "\\") {
									state = "eol-wsb";
								} else if (c === "#" && !qchar) {
									comment = true;
									rollback(S);
								}

								break;

							default:
								// list|command-flag
								// The following character after the initial
								// '$' must be a '('. If it does not follow,
								// error.
								//   --help=$"cat ~/files.text"
								//   --------^ Missing '(' after '$'.
								if (type === "command-flag") {
									if (
										N.value.value.length === 1 &&
										c !== "("
									) {
										error(S);
									}
								}

								// The following logic, is precursor validation
								// logic that ensures braces are balanced and
								// detects inline comment.
								if (p !== "\\") {
									if (c === "(" && !qchar) {
										braces.push(S.i);
									} else if (c === ")" && !qchar) {
										// If braces len is negative, opening
										// braces were never introduced so
										// current closing brace is invalid.
										if (!braces.length) {
											error(S);
										}
										braces.pop();
										if (!braces.length) {
											state = "eol-wsb";
										}
									}

									if (cin(C_QUOTES, c)) {
										if (!qchar) {
											qchar = c;
										} else if (qchar === c) {
											qchar = "";
										}
									}

									if (c === "#" && !qchar) {
										if (!braces.length) {
											comment = true;
											rollback(S);
										} else {
											S.column =
												braces.pop() -
												S.tables.linestarts[S.line];
											S.column++; // Add 1 to account for 0 base indexing.
											error(S);
										}
									}
								}
						}

						N.value.end = S.i;
						N.value.value += c;
					}
				}

				break;

			case "eol-wsb":
				if (cnotin(C_SPACES, c)) error(S);

				break;
		}
	}

	validate(S, N);
	add(S, N);
};
